util/pyclient/bft_client.py

```diff
@@ -241,3 +230,258 @@ def get_rsi_replies(self):
+            pass
+        # We got EOF or an exception - close the stream
+        if dest_addr in self.ssl_streams:
+            del self.ssl_streams[dest_addr]
+            self.establish_ssl_stream_parklot[dest_addr].unpark()
+            await stream.aclose()
+        return False
+
+    async def _receive_from_replica(self, dest_addr, replicas_addr, required_replies, cancel_scope):
+        """
+        Receive from a single replica. 3 stages:
+        1) Wait for the 1st 4 bytes which contains the header (payload length)
+        2) Receive the rest of the payload.
+        3) Process the received message payload.
+        If there is an error in stages 1 or 2 - exit straight (connection is closed inside _stream_recv_some)
+        """
+        if dest_addr not in self.ssl_streams:
+            self.establish_ssl_stream_parklot[dest_addr].unpark()
+            return
+        data = bytearray()
+        stream = self.ssl_streams[dest_addr]
+        while len(data) < self.MSG_HEADER_SIZE:
+            if not await self._stream_recv_some(data, dest_addr, stream, self.MSG_HEADER_SIZE):
+                return
+        payload_size = int.from_bytes(data[:self.MSG_HEADER_SIZE], "little")
+        del data[:self.MSG_HEADER_SIZE]
+        while len(data) < payload_size:
+            if not await self._stream_recv_some(data, dest_addr, stream, payload_size - len(data)):
+                return
+        try:
+            self._process_received_msg(bytes(data), dest_addr, replicas_addr, required_replies, cancel_scope)
+        except (bft_msgs.MsgError , struct.error) as ex:
+            # TCP is a stream protocol and we can receive in a certain period of time a broken stream of bytes
+            return
+
+    async def _recv_data(self, required_replies, dest_replicas, cancel_scope):
+        """
+        Receive reply messages until a quorum is achieved or the enclosing cancel_scope times out.
+        """
+        replicas_addr = [(r.ip, r.port) for r in dest_replicas]
+        async with trio.open_nursery() as nursery:
+            for dest_addr in replicas_addr:
+                if dest_addr in self.ssl_streams.keys():
+                    nursery.start_soon(self._receive_from_replica, dest_addr, replicas_addr, required_replies,
+                                       nursery.cancel_scope)
+                else:
+                    self.establish_ssl_stream_parklot[dest_addr].unpark()
+
+    async def __aenter__(self):
+        """async context manager method for 'with' statements"""
+        pass
+
+    async def __aexit__(self, *args):
+        """async context manager method for 'with' statements"""
+        self.exit_flag = True
+        for lot in self.establish_ssl_stream_parklot.values():
+            lot.unpark()
+        remove = [ k for k in self.ssl_streams ]
+        for k in remove:
+            stream = self.ssl_streams[k]
+            del self.ssl_streams[k]
+            await stream.aclose()
```

Comment:
instead of closing those sequentially, can we do this concurrently in a nursery?

Reply:
Good idea. I will.